README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# wcep

<!-- badges: start -->

<!-- badges: end -->

The goal of wcep is to analyze time-to-event data with multiple events,
for instance death and shock with related severity weights. The weight
of death is always 1 and other events have a positive weight smaller
than 1.

This package provides two main functions: wcep function produces
survival probabilities with related variances, and plot function uses
survival probabilities to illustrate Kaplan Meier curve with prespecified
confidence interval. Also, it provides a dataset.

## Installation

You can install the released version of wcep from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("wcep")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("nabipoor/wcep")
```

## Example

Basic examples provided in the Help document.


NEWS.md
# wcep 1.0.0

* Added a `NEWS.md` file to track changes to the package.

## Comments
This is the first version of this package. There is no news.




tests/testthat.R
library(testthat)
library(wcep)

test_check("wcep")


tests/testthat/testinputs.R
library(wcep)

test_that("testinputs",{

  expect_equal(wcep(data.frame(toyexample,1),data.frame(event=c('CHF','DTH','SHK','REMI'), weight = c(0.3,1,0.5,0.2)),
                    alpha = 0.05 , split = FALSE),
              noquote("Error: Data frame x should have 3 columns for one group or 4 columns for two groups comparison"))

  expect_equal(wcep(toyexample[,1:2],data.frame(event=c('CHF','DTH','SHK','REMI'), weight = c(0.3,1,0.5,0.2)),
                    alpha = 0.05 , split = FALSE),
               noquote("Error: Data frame x should have 3 columns for one group or 4 columns for two groups comparison"))

  expect_equal(wcep(toyexample,data.frame(event=c('CHF','DTH','SHK','REMI'),weight = c(0.3,1,0.5,0.2)), alpha = 1.05 , split = FALSE),
               noquote("Error: value of alpha should be between 0 and 1"))

  expect_equal(wcep(toyexample,data.frame(event=c('CHF','DTH','SHK','REMI'),weight = c(0.3,1,0.5,0.2)), alpha = -1.05 , split = FALS),
               noquote("Error: value of alpha should be between 0 and 1"))

  expect_equal(wcep(toyexample,data.frame(event=c('CHF','DTH','SHK','REMI'),weight = c(0.3,1,0.5,0.2)), alpha = 0 , split = FALSE),
               noquote("Error: value of alpha should be between 0 and 1"))

  expect_equal(wcep(toyexample[,1:3],data.frame(event=c('CHF','DTH','SHK','REMI'), weight = c(0.3,1,0.5,0.2)),
                    alpha = 0.05 , split = TRUE),
               noquote("Error: Data frame x should have 4 columns"))

  expect_equal(wcep(data.frame(toyexample[,1:3],sex=c(rep("M",40), rep("F",40), rep("G",24))),
                    data.frame(event=c('CHF','DTH','SHK','REMI'), weight = c(0.3,1,0.5,0.2)),
                    alpha = 0.05 , split = TRUE),
               noquote("Error: The last column should have two levels"))

  expect_equal(wcep(matrix(rep(1,40),ncol=4), data.frame(event=c('CHF','DTH','SHK','REMI'), weight = c(0.3,1,0.5,0.2)),
                    alpha = 0.05),
               noquote("Error: The second column should be factor"))
})
#


R/data.R
#' Toy example
#'
#' A data set containing patient IDs, event types, event times, and
#' gender of 100 patients
#'
#' @format A data frame with 104 rows and 4 columns
#' \describe{
#'    \item{PTID}{ID number of patients}
#'    \item{EvTp}{Event Types: SHK as Shock, CHF as Congestive Heart
#'          Failure, REMI as Recurrent Myocardial Infarction, DTH as
#'          Death; and N as No event}
#'    \item{EvTm}{Event Time (day)}
#'    \item{sex}{Gender of patients, M as Male, F as Female}
#'    }
#'  @source It is a generated example based on ASSENT-3.
#'  Armstrong PW, Westerhout CM, Van de Werf F, Califf RM, Welsh RC,
#'  Wilcox RG, Bakal JA
#'  Refining clinical trial composite outcomes: an application to the
#'  Assessment of the Safety and Efficacy of a New Thrombolytic-3
#'  (ASSENT-3) trial
#'  \url{https://pubmed.ncbi.nlm.nih.gov/21570513}
"toyexample"


R/plot.R
#' wcep plot
#'
#' Create a plot of Kaplan-Meier curve with its specified confidence iterval
#' @param x is an object of class "wcep"
#' @param main title of plot
#' @param type type of plot
#' @param lty line type
#' @param lwd line width
#' @param xlab first axis lable
#' @param ylab second axis lable
#' @param xlim first axis limits
#' @param ylim second axis limits
#' @param cex legend font size
#' @param ... other parameters of generic "plot" have no use here
#' setOldClass("wcep")
#' @export


  plot.wcep <- function(x, main = " ", type = "n", lty = NULL,  lwd = NULL,
                          xlab = " ", ylab = "Survival Probability", xlim = NULL,
                          ylim = NULL, cex=NULL, ...){
      if(length(x) == 5) {
        timelab <- 0:dim(x$life_table)[2]
        ymin <- round(max(min (x$lower)-0.0499,0)-0.05, 1)
        yl <- if (is.null(ylim)) c(ymin, 1) else ylim
        xmax <- length(x$survival_probabilities)
        xl <- if (is.null(xlim)) c(0, xmax) else xlim
        typ <- if (type == "n") "l" else type
        lt <- if (is.null(lty)) 1 else lty
        lw <- if (is.null(lwd)) 2 else lwd
        plot(timelab, c(1,x$survival_probabilities), type = typ, lwd = lw, col = "hotpink2",
             main = main, xlab = xlab, ylab = ylab, xlim = xl, ylim = yl, ...)
        b <- c(c(1, x$upper), rev(c(1, x$lower)))
        c <- c(timelab,rev(timelab))
        polygon(c, b, col = "lavenderblush1", border = NA)
        points(timelab, c(1,x$survival_probabilities), type = typ, lty = lt, lwd = lw,
               col = "hotpink2")
      } else {
        timelab <- 0:dim(x[[1]]$life_table)[2]
        ymin <- round(max(min (c(x[[1]]$lower, x[[2]]$lower))-0.0499,0)-0.05, 1)
        yl <- if (is.null(ylim)) c(ymin,1) else ylim
        xl <- if (is.null(xlim)) c(0, length(x[[1]]$survival_probabilities)) else xlim
        typ <- if (type == "n") "l" else type
        lt <- if (is.null(lty)) 1 else lty
        lw <- if (is.null(lwd)) 2 else lwd
        plot(timelab, c(1,x[[1]]$survival_probabilities), type = typ, lwd = lw, col = "violetred4",
             main = main, xlab = xlab, ylab = ylab, xlim = xl, ylim = yl, lty = lt, ...)
        b1 <- c(c(1, x[[1]]$upper), rev(c(1, x[[1]]$lower)))
        c1 <- c(timelab,rev(timelab))
        polygon(c1, b1, col = rgb(0.92, 0.43, 0.63, 0.43), border = NA)

        b2 <- c(c(1, x[[2]]$upper), rev(c(1, x[[2]]$lower)))
        c2 <- c(timelab,rev(timelab))
        polygon(c2, b2, col = rgb(1,0.58,0.55,0.62), border = NA)
        points(timelab, c(1,x[[2]]$survival_probabilities), type = typ, lty = lt,
               lwd = lw, col = "orangered1")
        cx <- if (is.null(cex)) 0.8 else cex
        legend("bottomleft", legend = names(x)[1:2], lty = rep(lt, 2), lwd = rep(lw, 2),
               col = c("violetred4", "orangered1"), cex=rep(cx,2))
      }
    }


R/wcep_core.R
#' @include nam.R
#'
# Function: wcep_core
#
# Analyze Weighted Composite EndPoints for one data set and produce life table, survival
# probabilities, variances, and 95% C.I. by given alpha or by default alpha = 0.05.
#
# Authors: Majid Nabipoor, Jeff Bakal
# revised: Sep. 2019, Jan. 2020, Oct. 2020

 wcep_core <- function(x, ew, alpha) {

              # list of output
              out <- list()
              class(out) <- "wcep"
              # add a numerical patient ID column
              pt_h <- as.factor(x[, 1])
              ptid_h <- 1:length(unique(pt_h))
              dd <- data.frame(pt_h = unique(pt_h), ptid_h)
              names(dd) <- c(names(x)[1], "c4")
              x1 <- merge(x, dd, by = names(x)[1])

              # Numerate similar events for a patient; SHK SHK -> SHK1 SHK2
              evtp_h <- as.character(x1[, 2])
              evtm_h <- as.numeric(x1[, 3])
              invisible(sapply(ptid_h, function(i) {ind <- which(x1[, 4] == i);
              evtp_h[ind] <<- nam(evtp_h[ind])}))
              x2 <- data.frame(c1 = x1[, 4], evtp_h, evtm_h = x1[, 3])

              # time matrix for each unique event of patients
              xx <- x2 %>% spread(evtp_h, evtm_h)
              maxtime <- max(x2[, 3])
              xx[is.na(xx)] <- maxtime+1

              # life table and survival probabilities
              names(ew)<-c("event","weight")
              ew1 <- data.frame(event = ew[, 1], weight=ew[, 2])
              ew_h <- data.frame(cbind(colnames(xx[, -1]), substr(colnames(xx[, -1]), 1,
                      nchar(colnames(xx[, -1])) - 1)))
              colnames(ew_h) <- c("c1", "event")
              ne_h <- setdiff(ew_h[, 2], ew[, 1])
              ew_h0<- droplevels(ew_h[-which(ew_h$event==ne_h),],exclude=ne_h)
              ew_h1 <- left_join(ew_h0, ew1, by="event", stringsAsFactors=FALSE)[, c(1, 3)]
              xx <- xx[,-which(colnames(xx) %in% setdiff(colnames(xx), ew_h1[,1]))]
              n <- length(ptid_h)
              lifemat <- matrix(1, nrow = n, ncol = maxtime+1)
              tot <- n + maxtime
              pb <- progress_bar$new(
                format = "  downloading :Progress [:bar] :percent",
                clear = FALSE, total = tot, width = 80)
              s_table <- sapply(1:n, function(i){
                pb$tick()
                sapply(2:(maxtime + 1), function(j){
                lifemat[i,j] <<- prod(apply(cbind(t(xx[i, ]), ew_h1)[, -2], 1, function(v){
                                 ifelse(v[1] <= j-1, 1-as.numeric(v[2]), 1)}))
                                 })})
             # lt <- dim(s_table)[1]
             # s_table <- s_table[-lt, ]
              out$life_table <- t(s_table)
              out$survival_probabilities <- apply(s_table, 1, mean)

              # variance
              s_table1 <- data.frame(1, t(s_table))
              Key <- data.frame(key = 1 - ew$weight, ev = ew$event)
              s <- cumsum(sapply(2:dim(s_table1)[2], function(j){
                pb$tick()
                drev <- data.frame( key = s_table1[, j] / s_table1[, j - 1])
                f<-table(dplyr::left_join(drev, Key, by = "key")[, 2])
                u=attr(f,"dimnames")[[1]]
                p<-data.frame(key=Key[,2],0)
                p[,2]<-sapply(1:dim(p)[1],function(i){ifelse(length(which(u==p[i,1]))!=0, f[which(u==p[i,1])], 0)})
                pj<-matrix(p[,2] / sum(s_table1[, j - 1]), ncol=1)
                uj<- matrix(t(ew[, 2]), nrow=1)%*%pj
                (1 - uj)^(-2)*(t(ew[,2])%*%(diag(as.vector(pj))- pj%*%t(pj))%*%ew[,2])
              }))*apply(s_table, 1, mean)^2 / (dim(s_table)[2])
              out$variance <- s
              upper <- out$survival_probabilities +  qnorm(1-alpha / 2) * sqrt(s)
              lower <- out$survival_probabilities -  qnorm(1-alpha / 2) * sqrt(s)
              out$upper <- sapply(1:length(upper), function(i) {
                ifelse(upper[i] > 1, upper[i] <- 1, upper[i])
                ifelse(upper[i] < 0, upper[i] <- 0, upper[i])
                })
              out$lower <- sapply(1:length(lower), function(i) {
                ifelse(lower[i] > 1, lower[i] <- 1, lower[i])
                ifelse(lower[i] < 0, lower[i] <- 0, lower[i])
              })
              out
         }


R/wcep.R
#' @include nam.R wcep_core.R
NULL

#'  Analysis of weighted composite endpoints
#'
#'  Analyze given data frame and return Kaplan-Meier survival probabilities together with the specified confidence interval.
#'  \code{wcep} modifies Kaplan-Meier curve by taking into account severity weights of different event. Alternative methods are Anderson Gill model and win ratio of composite outcomes.The function takes event dataset and user-specified severity weights  to generate a modified Kaplan-Meier curve and comparison statistics based on the weighted composite endpoint method.  The user supplies the event data set, the weights, and the factor to split on . The package will generate the weighted survival curve, confidence interval and test the differences between the two groups.
#'
#' @param x This data frame usually has 3 columns. The first column specifies patient ID,
#' which is a character or numeric vector, the second column is a factor with character values
#' of event types. The third column is a numeric vector of event times. If split = TRUE,
#' then the forth column is a character vector of split groups of at most two groups,
#' like gender.
#'
#' @param EW This data frame has two columns. The first column
#' specifies a character vector of event types. The second column specify weights.
#' The naming of event types in x and EW should be exactly similar.
#'
#' @param alpha A numeric value between 0-1 which specifies the confidence level,
#' if it is not specified, by default is 0.05.
#'
#' @param split A logical value of T or F which allows to compare two groups.
#'
#' @section References:
#' Bakal J., Westerhout C. M., Armstrong P. W. (2015) Impact of weighted composite
#' compared to traditional composite endpoints for the design of randomized
#' controlled trails, Statistical Methods in Medicine Research Vol. 24(6) 980-988
#' @examples
#' data(toyexample)
#' #event weights
#' EW <- data.frame(event = c('CHF','DTH','SHK','REMI'), weight = c(0.3,1,0.5,0.2))
#' res1 <- wcep(toyexample, EW)
#' str(res1)
#' res1$survival_probabilities
#' plot(res1)
#' #comparing two genders
#' res2 <- wcep(toyexample, EW, split=TRUE)
#' plot(res2)
#' #wilcox and t test
#' res2$Wilcoxontest
#' res2$t_test
#' @author
#' Majid Nabipoor: nabipoor@@ualberta.ca,
#' Cynthia Westerhout: cindy.westerhout@@ualberta.ca,
#' Jeffrey Bakal: jbakal@@ualberta.ca
#' @seealso \code{\link{coxph}} for Anderson Gill model
#' @importFrom stats qnorm t.test
#' @importFrom graphics plot points polygon legend
#' @importFrom grDevices rgb
#' @import coin dplyr progress tidyr
#' @export

 wcep <- function(x, EW, alpha = 0.05 , split = FALSE){

          if (dim(x)[2] < 3 | dim(x)[2] > 4) {
              return(noquote("Error: Data frame x should have 3 columns for one group or 4 columns for two groups comparison"))
          }
          if (alpha >= 1 | alpha <= 0) {
              return(noquote("Error: value of alpha should be between 0 and 1"))
          }
          if ( split == TRUE && dim(x)[2] != 4 ) {
              return(noquote("Error: Data frame x should have 4 columns" ))
          }
          if ( split == TRUE && length(unique(x[,4])) > 2 ) {
              return(noquote("Error: The last column should have two levels" ))
          }
          if ( is.factor(x[,2]) == FALSE ) {
              return(noquote("Error: The second column should be factor" ))
          }
          res <- structure(list(), class = "wcep")
          if(split == FALSE) {
            pb <- progress_bar$new(
              format = " work progress [:bar] :percent",
              total = NA, clear = FALSE, width= 80)
            res <- wcep_core(x[, 1:3], EW, alpha)
          } else {
            groups <- unique(x[, 4])
            for(i in 1:2) {
              pb <- progress_bar$new(
                format = " Progress [:bar] :percent",
                total = NA, clear = FALSE, width= 80)
              res[[paste0(" ", groups[i], sep="")]] <- wcep_core(x[which(x[, 4] == groups[i]), 1:3], EW,                                                                          alpha)
            }
            res$Wilcoxontest <- wilcoxsign_test((res[[paste0(" ", groups[1], sep="")]])$survival_probabilities ~
                                (res[[paste0(" ", groups[2], sep="")]])$survival_probabilities, zero.method = c("Pratt"))
            res$t_test <- (t.test((res[[paste0(" ", groups[1], sep="")]])$survival_probabilities,
                          (res[[paste0(" ", groups[2], sep="")]])$survival_probabilities))
          }
          res
 }
